x-pack/plugins/monitoring_collection/server/plugin.ts

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { JsonObject } from '@kbn/utility-types';
+import { CoreSetup, Plugin, PluginInitializerContext, Logger } from 'kibana/server';
+import { registerDynamicRoute } from './routes';
+import { MakeSchemaFrom } from '../../../../src/plugins/usage_collection/server';
+import { TYPE_ALLOWLIST } from './constants';
+
+export interface MonitoringCollectionSetup {
+  registerMetric: <T>(metric: Metric<T>) => void;
+}
+
+export type MetricResult<T> = T & JsonObject;
+
+export interface Metric<T> {
+  type: string;
+  schema: MakeSchemaFrom<T>;
+  fetch: () => Promise<MetricResult<T> | Array<MetricResult<T>>>;
+}
+
+export class MonitoringCollectionPlugin implements Plugin<MonitoringCollectionSetup, void, {}, {}> {
+  private readonly initializerContext: PluginInitializerContext;
+  private readonly logger: Logger;
+
+  private metrics: Record<string, Metric<any>> = {};
+
+  constructor(initializerContext: PluginInitializerContext) {
+    this.initializerContext = initializerContext;
+    this.logger = initializerContext.logger.get();
+  }
+
+  async getMetric(type: string) {
+    if (this.metrics.hasOwnProperty(type)) {
+      return await this.metrics[type].fetch();
```

Comment:
Can there be a requirement that the fetch methods are synchronous? Internally, the plugins can manage a state object that is continually updated. Then their fetch methods would just return a copy of that.

Reply:
Maybe the downside is, if a plugin needs to use Elasticsearch to get its stats, then it would need to keep querying those in the background in order to have its stats ready -- but it doesn't know if metricbeat will ever use those stats.